### DISCUSSION TITLE
ISG behind a reverse proxy that rewrites the URI

### DISCUSSION THREAD

**Discussion by fernandobandeira** [vercel/next.js#17581] [Help] [answered]
# Bug report

## Describe the bug

This bug originated from discussion https://github.com/vercel/next.js/discussions/17260

We were exploring ways to use ISG on a multi-tenant app and thought that putting all pages inside a [hostname] param would do the trick, then we could have a reverse proxy on top of next.js to hide the [hostname] from the client URL.

There are 2 bugs with this approach

1 - Client-side routing is broken:
The client-side routing expects the hostname to be part of the URL and since we're hiding it, the user can't navigate between the pages anymore, what happens is that if the user tries to go to a page like /example, an example would be treated as the hostname and he'd be taken to the /[hostname] route instead of /[hostname]/example.

2 - The first generation doesn't work
The first static generation isn't triggered when using the reverse proxy so all of the requests see the fallback page, interesting enough if the page was generated previously or from another means it'll serve it just fine, it also regenerates the existing static page fine too, it's just the first generation that doesn't occur.

## To Reproduce

You can test both bugs with this repo:

https://github.com/fernandobandeira/isg-next-poc-multi-tenant

There are reproduction steps there and I tried to make it really simple to set up and test the issues, please ping me if you have problems.

## Expected behavior

I think it'd be nice to have this sorted out for multi-tenant apps, it'd be really great, although I'm not sure how many users would benefit from it, I imagine multi-tenant apps are common but maybe they aren't.

Another idea would be to add some built-in functionality to group the generated pages by the hostname that doesn't require the proxy. although I'm not sure that'd be possible since the generation is done without the request details apart from the path but maybe that could be builtin somehow.

**Comment by calirvine** [comment]
I've spent the last couple of days experimenting trying to find a way to accomplish pretty much exactly this, I'd be really interested to see this feature.

**Reply by nicosh** [reply]
@fernandobandeira @calirvine you made any progress to accomplish this?

**Reply by steven-tey** [reply]
@calirvine @nicosh I just posted an example for this in my comment below, hope it helps!

**Comment by steven-tey** [comment]
Hey @fernandobandeira! Great question – we recently rolled out a [simple example](https://github.com/vercel/examples/tree/main/edge-functions/hostname-rewrites) that allows you to create unique content pages with a multi-tenant infrastructure using [edge function](https://vercel.com/features/edge-functions) rewrites and [ISR](https://vercel.com/docs/concepts/next.js/incremental-static-regeneration). Each of your users gets assigned a unique subdomain when they create their account, with the (usually paid) option to add a custom domain. 

We're also rolling out more complex boilerplates (e-commerce storefront builders, website builders) + supplementary code snippets (static tweets, domains API, OG image generator) in an [early access program](https://eap.vercel.sh/) – feel free to join if you're interested (click "sign up" on the Platforms on Vercel program).

**Reply by akinwol** [reply]
@steven-tey Just curious, any updates on the e-commerce/storefront builders boilerplates?  I signed up but just wondering if its available.

**Comment by steven-tey** [comment, marked as answer]
Quick update on this – we recently launched the Platforms Starter Kit that's a comprehensive template for building multi-tenant applications with built-in custom domains support.: https://demo.vercel.pub/platforms-starter-kit

Feel free to try it out and let us know what you think! :)
